### ISSUE TITLE
ActiveRecord memoization slowdown

### ISSUE THREAD

**Issue by amedrz** [jnunemaker/flipper#653]
We've been debugging an increase of about 40ms in our Rails application's P95 (prod) after installing `flipper` and `flipper-active_record`, even before making any use of it i.e. with an empty DB. The P95 went back to normal after removing `flipper-active_record`.

Doing some profiling in a local environment, we noticed the overhead might be caused from calls to [`get_all`](https://github.com/jnunemaker/flipper/blob/928a8d8731da6656bf3802abbce59eb2f8f078c2/lib/flipper/adapters/active_record.rb#L108-L123) on every request.

Here's some sampling of 2 consecutive requests:

### Request 1

<img width="1504" alt="03-24-bmp2n-mcwy9" src="https://user-images.githubusercontent.com/15628/182710646-b107468a-744e-4a61-8c1d-e80b641080f3.png">

### Request 2

<img width="1501" alt="03-25-7wpnr-j7qe2" src="https://user-images.githubusercontent.com/15628/182711965-fbc33433-63da-4468-aea8-52286e54804c.png">

In the 1st one it seems to do the full DB round, and use `ActiveRecord` cache in the 2nd. Perhaps that's outside of Flipper's control but I'm also wondering if the `Memoizer` works at at request level or across many?

Lastly, is the performance penalty we're seeing known and expected?

Thank you

**Comment by jnunemaker** [comment]
Hi @amedrz! 

Somehow missed this. Apologies!

This is likely the [preloading](https://www.flippercloud.io/docs/optimization#preloading), not the memoization. You can turn it off (see docs I just linked) and deploy again and see if that has the same affect as removing. 

Memoizing ensures that the same feature is never queried for more than once in a request. 

Preloading tries to load all the features (and gate data) as efficiently as flipper can (currently). At GitHub, we fronted an AR adapter with memcached and did not use preloading. 

The right approach is all dependent on your app's setup and perf characteristics. Happy to help more here or via email if you want or need more help but don't want anything to be public for any reason. 

So, to summarize: 

* turn off preloading
* deploy
* see if that avoids those queries, but note that now each feature will be loaded on demand with N+1 calls where N is the unique number of features checked for a given request

If you aren't happy with the performance of preloading or lazy loading, you can always use a cache adapter like github.com does/did (assuming those calls are faster on your infra) or there are more options like tiered approaches (in memory => memcache => AR).

**Comment by jnunemaker** [comment]
Closing to keep tidy but happy to help until we get you where you need to go.
